src/Subject.ts

```diff
@@ -2,236 +2,145 @@ import {Operator} from './Operator';
-    return new Subject<T>(destination, source);
-  };
+  isUnsubscribed = false;
 
-  constructor(protected destination?: Observer<T>, protected source?: Observable<T>) {
+  isStopped = false;
+
+  hasError = false;
+
+  thrownError: any = null;
+
+  constructor() {
     super();
-    this.source = source;
   }
 
-  public observers: Observer<T>[] = [];
-  public isUnsubscribed: boolean = false;
-
-  protected isStopped: boolean = false;
-  protected hasErrored: boolean = false;
-  protected errorValue: any;
-  protected dispatching: boolean = false;
-  protected hasCompleted: boolean = false;
+  static create: Function = <T>(destination: Observer<T>, source: Observable<T>): AnonymousSubject<T> => {
+    return new AnonymousSubject<T>(destination, source);
+  };
 
   lift<T, R>(operator: Operator<T, R>): Observable<T> {
-    const subject = new Subject(this.destination || this, this);
+    const subject = new AnonymousSubject(this, this);
     subject.operator = operator;
     return <any>subject;
   }
 
-  add(subscription: TeardownLogic): Subscription {
-    return Subscription.prototype.add.call(this, subscription);
-  }
-
-  remove(subscription: Subscription): void {
-    Subscription.prototype.remove.call(this, subscription);
-  }
-
-  unsubscribe(): void {
-    Subscription.prototype.unsubscribe.call(this);
-  }
-
-  protected _subscribe(subscriber: Subscriber<T>): TeardownLogic {
-    if (this.source) {
-      return this.source.subscribe(subscriber);
-    } else {
-      if (subscriber.isUnsubscribed) {
-        return;
-      } else if (this.hasErrored) {
-        return subscriber.error(this.errorValue);
-      } else if (this.hasCompleted) {
-        return subscriber.complete();
+  next(value: T) {
+    if (this.isUnsubscribed) {
+      throw new ObjectUnsubscribedError();
+    }
+    if (!this.isStopped) {
+      const { observers } = this;
+      const len = observers.length;
+      const copy = observers.slice();
+      for (let i = 0; i < len; i++) {
+        copy[i].next(value);
       }
-
-      this.throwIfUnsubscribed();
-
-      const subscription = new SubjectSubscription(this, subscriber);
-
-      this.observers.push(subscriber);
-
-      return subscription;
     }
   }
 
-  protected _unsubscribe(): void {
-    this.source = null;
-    this.isStopped = true;
-    this.observers = null;
-    this.destination = null;
-  }
-
-  next(value: T): void {
-    this.throwIfUnsubscribed();
-
-    if (this.isStopped) {
-      return;
+  error(err: any) {
+    if (this.isUnsubscribed) {
+      throw new ObjectUnsubscribedError();
     }
-
-    this.dispatching = true;
-    this._next(value);
-    this.dispatching = false;
-
-    if (this.hasErrored) {
-      this._error(this.errorValue);
-    } else if (this.hasCompleted) {
-      this._complete();
+    this.hasError = true;
+    this.thrownError = err;
+    this.isStopped = true;
+    const { observers } = this;
```

Comment:
It seems we can get away without copying `observers` both here and in `complete`. After `hasError` or `isStopped` is true, all new applicants to `subscribe` con't make it into the observers list. Additionally, setting `this.observers` to null before notifying will save us the cost of each `SubjectSubscription` individually slicing their observer out of the observers list.